### PR TITLE
Deprecate configure on demand

### PR DESCRIPTION
**Pull request by oehme**
This feature was introduced to make configuration time in
large projects acceptable. This was back at a time when Gradle
was relatively inefficient. By now, the feature has lost a lot
of its appeal, due to many performance improvements.

Apart from the reduced usefulnes, it was also very hard to use
correctly and confused many users. E.g. the use of `allprojects {}`
 or `subprojects {}` negates the benefit of CoD, yet this is the
canonical way to cross-configure projects.

Going forward, we will ask users to provide profiling results if they
are experiencing slow configuration times. In most cases the problems
could be fixed in the user's build itself. In some cases we can make
some transparent improvement to Gradle.

### CODE REVIEW

File: subprojects/cli/src/main/java/org/gradle/cli/CommandLineParser.java [hunk 1]
```diff
@@ -454,18 +454,14 @@ public ParserState onComplete() {
...
-            
+
             ParsedCommandLineOption parsedOption = commandLine.addOption(optionString.option, option);
             if (values.size() + parsedOption.getValues().size() > 1 && !option.getAllowsMultipleArguments()) {
                 throw new CommandLineArgumentException(String.format("Multiple arguments were provided for command-line option '%s'.", optionString));
             }
             for (String value : values) {
                 parsedOption.addArgument(value);
             }
-            if (option.getDeprecationWarning() != null) {
```
Comment (blindpirate) [inline subprojects/cli/src/main/java/org/gradle/cli/CommandLineParser.java]: Why is this removed? If so, the deprecation message would no longer make sense. https://github.com/gradle/gradle/blob/8c091930436481fc73bc4bf6c9ab7b397b31b60b/subprojects/build-option/src/main/java/org/gradle/internal/buildoption/CommandLineOptionConfiguration.java#L94

File: subprojects/core/src/integTest/groovy/org/gradle/api/ConfigurationOnDemandIntegrationTest.groovy [hunk 2]
```diff
@@ -33,32 +33,6 @@ class ConfigurationOnDemandIntegrationTest extends AbstractIntegrationSpec {
...
-        when:
-        run("foo", "--configure-on-demand", "--parallel")
-
-        then:
-        fixture.assertProjectsConfigured(":")
-        output.count("Parallel execution with configuration on demand is an incubating feature") == 1
-    }
-
     def "can be enabled from command line for a single module build"() {
         file("gradle.properties") << "org.gradle.configureondemand=false"
```
Comment (blindpirate) [inline subprojects/core/src/integTest/groovy/org/gradle/api/ConfigurationOnDemandIntegrationTest.groovy]: Why there's no `expectDeprecation()` in this test?

File: subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/StaleOutputHistoryLossIntegrationTest.groovy [hunk 3]
```diff
@@ -235,6 +235,9 @@ class StaleOutputHistoryLossIntegrationTest extends AbstractIntegrationSpec {
         javaProjects.each {
             forceDelete(it.redundantSourceFile)
         }
+        if (expectDeprecationWarning) {
```
Comment (blindpirate) [inline subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/StaleOutputHistoryLossIntegrationTest.groovy]: Maybe we can write `if(arguments.contains('--configure-on-demand')){...}`?

File: subprojects/platform-native/src/integTest/groovy/org/gradle/nativeplatform/LibraryDependenciesIntegrationTest.groovy [hunk 4]
```diff
@@ -248,7 +248,7 @@ project(":lib") {
 """
 
         when:
-        executer.withArgument('--configure-on-demand')
+        executer.withArgument('--configure-on-demand').expectDeprecationWarning()
```
Comment (blindpirate) [inline subprojects/platform-native/src/integTest/groovy/org/gradle/nativeplatform/LibraryDependenciesIntegrationTest.groovy]: Is it possible to put all `expectDeprecationWarning` in one place? Like inside `withArguments` methods or `run` method. Of course, this may be not feasible because of various complex situation, just a discussion.

File: subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/StartParamsValidatingActionExecuter.java [hunk 5]
```diff
@@ -60,6 +62,12 @@ public Object execute(BuildAction action, BuildRequestContext requestContext, Bu
                 throw new IllegalArgumentException(String.format("The specified settings file '%s' is not a file.", startParameter.getSettingsFile()));
             }
         }
+        if (startParameter.isConfigureOnDemand()) {
```
Comment (blindpirate) [inline subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/StartParamsValidatingActionExecuter.java]: Still, this is a conflict with intention of #2890  . Personally, I prefer #2890 's way, putting all deprecation message together to avoid writing `nagUserWith` again and again.

File: subprojects/cli/src/main/java/org/gradle/cli/CommandLineParser.java (hunk 1 above)
Comment (oehme) [reply subprojects/cli/src/main/java/org/gradle/cli/CommandLineParser.java]: It does still make sense, because it is shown in the `gradle help` output.

I removed this log message in this place, because it was not the right place for several reasons:

- it could be logged before our logging infrastructure was initialized, meaning the message got lost in embedded mode
- it was only logged when used with the CLI switch, but not logged when used with the Gradle property

File: subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/StaleOutputHistoryLossIntegrationTest.groovy (hunk 3 above)
Comment (oehme) [reply subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/StaleOutputHistoryLossIntegrationTest.groovy]: I didn't want to repeat myself here, so I made it part of the `where:` table instead.

File: subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/StartParamsValidatingActionExecuter.java (hunk 5 above)
Comment (oehme) [reply subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/StartParamsValidatingActionExecuter.java]: It only printed them when activated through the CLI and it printed them too early.

We can open a follow-up issue to truly centralize this. That solution would have to
- work with CLI and Gradle properties
- collect messages and only log them after the logging system is active

File: subprojects/platform-native/src/integTest/groovy/org/gradle/nativeplatform/LibraryDependenciesIntegrationTest.groovy (hunk 4 above)
Comment (oehme) [reply subprojects/platform-native/src/integTest/groovy/org/gradle/nativeplatform/LibraryDependenciesIntegrationTest.groovy]: Unfortunately, the expectations is reset after every call to `run`. So I only have two options:

- add it to every `run` call
- override `run` for the whole class (as I did in the other test)

Unless of course I'm missing some feature in that test fixture. Maybe @big-guy has an idea?

File: subprojects/cli/src/main/java/org/gradle/cli/CommandLineParser.java (hunk 1 above)
Comment (blindpirate) [reply subprojects/cli/src/main/java/org/gradle/cli/CommandLineParser.java]: Maybe we could remove the unused `deprecationPrinter` property and related constructor of `CommandLineParser` class?

File: subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/StartParamsValidatingActionExecuter.java (hunk 5 above)
Comment (blindpirate) [reply subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/StartParamsValidatingActionExecuter.java]: Missing `@SuppressWarnings("deprecation")` on this method.

File: subprojects/core/src/main/java/org/gradle/configuration/DefaultBuildConfigurer.java [hunk 6]
```diff
@@ -19,11 +19,8 @@
...
 import org.gradle.execution.ProjectConfigurer;
 import org.gradle.util.SingleMessageLogger;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
 
 public class DefaultBuildConfigurer implements BuildConfigurer {
-    private static final Logger LOGGER = LoggerFactory.getLogger(DefaultBuildConfigurer.class);
     private final ProjectConfigurer projectConfigurer;
 
     public DefaultBuildConfigurer(ProjectConfigurer projectConfigurer) {
```
Comment (blindpirate) [inline subprojects/core/src/main/java/org/gradle/configuration/DefaultBuildConfigurer.java]: `@SuppressWarning` here.

File: subprojects/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/BuildModelActionRunner.java [hunk 7]
```diff
@@ -37,6 +37,7 @@
 
 public class BuildModelActionRunner implements BuildActionRunner {
     @Override
+    @SuppressWarnings("deprecation")
```
Comment (blindpirate) [inline subprojects/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/BuildModelActionRunner.java]: Looks like not all invocations of  `setConfigureOnDemand` and `isConfigureOnDemand` are annotated with `SuppressWarnings`:

https://github.com/gradle/gradle/blob/a784b7511b82bd8acacc4c4751fd85a6de8734fe/subprojects/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/ClientProvidedBuildActionRunner.java#L47

https://github.com/gradle/gradle/blob/f727cfb1a8edc6564ec0556be8d41e9c7d92718a/subprojects/composite-builds/src/main/java/org/gradle/composite/internal/DefaultIncludedBuildFactory.java#L112

https://github.com/gradle/gradle/blob/a784b7511b82bd8acacc4c4751fd85a6de8734fe/subprojects/core/src/main/java/org/gradle/initialization/StartParameterBuildOptionFactory.java#L251

https://github.com/gradle/gradle/blob/f2af2a3f0dc03ec0be205c5689589591bc4fb23f/subprojects/core/src/main/java/org/gradle/initialization/DefaultGradleLauncher.java#L325

File: subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/StartParamsValidatingActionExecuter.java (hunk 5 above)
Comment (blindpirate) [reply subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/StartParamsValidatingActionExecuter.java]: With these questions, we should open a new issue to address it. Would it be a good idea if we put each `nagUserWith` in each `BuildOption`'s `applyTo` method?

File: subprojects/cli/src/main/java/org/gradle/cli/CommandLineParser.java (hunk 1 above)
Comment (oehme) [reply subprojects/cli/src/main/java/org/gradle/cli/CommandLineParser.java]: Yes, you're right I missed that.

File: subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/StartParamsValidatingActionExecuter.java (hunk 5 above)
Comment (oehme) [reply subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/StartParamsValidatingActionExecuter.java]: Possibly, I haven't looked deeply at it. Let's just open an issue describing the problem and we can figure out the implementation later.

File: subprojects/core/src/main/java/org/gradle/configuration/DefaultBuildConfigurer.java (hunk 6 above)
Comment (oehme) [reply subprojects/core/src/main/java/org/gradle/configuration/DefaultBuildConfigurer.java]: SuppressWarnings is only needed in projects using strict compilation (e.g. the tooling API). You don't need to check for this in a review, the compiler does it already.

File: subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/StartParamsValidatingActionExecuter.java (hunk 5 above)
Comment (blindpirate) [reply subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/StartParamsValidatingActionExecuter.java]: Sure.

File: subprojects/core/src/main/java/org/gradle/configuration/DefaultBuildConfigurer.java (hunk 6 above)
Comment (blindpirate) [reply subprojects/core/src/main/java/org/gradle/configuration/DefaultBuildConfigurer.java]: Good to know... so does this mean that we don't care about the deprecation warning outputted in gradle compile tasks?

Comment (oehme) [reply subprojects/core/src/main/java/org/gradle/configuration/DefaultBuildConfigurer.java]: I think we haven't cared much so far, but I might be mistaken. Maybe we should just standardize on strict compilation. That would certainly make the logs shorter by forcing people to think about every warning.

**Comment by oehme** [comment]
We decided not to do this for 4.3, but solicit more feedback from our user base first.
